pkg/apis/wsecurity/v1/count.go

```diff
@@ -0,0 +1,146 @@
+package v1
+
+import (
+	"bytes"
+	"fmt"
+)
+
+//////////////////// CountProfile ////////////////
+
+// Exposes ValueProfile interface
+type CountProfile uint8
+
+func (profile *CountProfile) Profile(args ...interface{}) {
```

Comment:
In particular, this code will `panic` if the `args[0]` is not a `uint8`, and ignore subsequent arguments.

The interface here feels like it's actively fighting against Go's type system to make it something more like Javascript. I'm not sure I understand what the value of this `Profile` interface is.

Reply:
This is an unfortunate truth...

The previous version of SG does not use a central interface for the dataTypes. Instead, there are some 15 data types, which all were designed with very similar methods (such as Decide, Learn, Merge, Add etc.). Each dataType uses a different method signature - for example:
```golang
func (p *KeyValPile) Add(kv *KeyValProfile) {...}
func (p *SimpleValPile) Add(svp *SimpleValProfile) {...}
func (p *UrlPile) Add(u *UrlProfile) {...}
```

I refactored the code in this version to use a central interface   - i.e. `v1.go`

This aligns all data types to a single interface which has many advantages in my mind. The obvious disadvantage is what you bring up here - it moves the data type assertion from compile time to run time. I guess we can decide to avoid a central interface and go with the code as is from the previous version - but I am not sure this is a better path.

Reply:
I think this is where type constraints with generics could be helpful -- You might be able to declare an interface like:

```golang
type Pile[T any] interface {
  Add(Profile[T])
  Clear()
  fmt.Stringer
}
```

Reply:
It seems as indicated above that generics will not change the number of core dataTypes we need to support since their implementation is quite different from one another, learning an HttpHeader is very different from learning a Counter or learning a URL. 
Generics will make each core dataType become more generic which may become useful at some point.